src/cursor/cur_hs.c

```diff
@@ -589,19 +589,25 @@ __wt_curhs_search_near_after(WT_SESSION_IMPL *session, WT_CURSOR *cursor)
 static int
 __curhs_search_near_helper(WT_SESSION_IMPL *session, WT_CURSOR *cursor, bool before)
 {
+    WT_CURSOR_HS *hs_cursor;
     WT_DECL_ITEM(srch_key);
     WT_DECL_RET;
     int cmp;
 
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+
     WT_RET(__wt_scr_alloc(session, 0, &srch_key));
     WT_ERR(__wt_buf_set(session, srch_key, cursor->key.data, cursor->key.size));
     WT_ERR(cursor->search_near(cursor, &cmp));
     if (before) {
         /*
          * If we want to land on a key that is smaller or equal to the specified key, keep walking
-         * backwards as there may be content inserted concurrently.
+         * backwards as there may be content inserted concurrently. Note that there is an
+         * interesting edge case when the history cursor only sets the btree ID as it's key range. A
+         * record can be concurrently inserted into history store that points before the positioned
+         * cursor.
          */
-        if (cmp > 0) {
+        if (F_ISSET(hs_cursor, WT_HS_CUR_KEY_SET) && cmp > 0) {
```

Comment:
This change may cause problems, Instead of checking whether the key is set or not? How about comparing the btreeid in the if {} check and similar to the other code paths where we exit.